src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/model/Execution.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * pottery-backend-interface - Backend API for testing programming exercises
+ * Copyright © 2015-2018 BlueOptima Limited, Andrew Rice (acr31@cam.ac.uk)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package uk.ac.cam.cl.dtg.teaching.pottery.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.wordnik.swagger.annotations.ApiModelProperty;
+
+/**
+ * An Execution represents a program that can be run by Pottery under a particular image.
+ *
+ * The program command-line can contain variables delimited with @'s (like @VARIABLE@) which are
+ * replaced by Pottery according to the context the Execution is run in.
+ *
+ * The container that is started has restrictions applied to it as specified; if none are specified,
+ * Pottery can apply default restrictions.
+ */
+public class Execution {
+  @ApiModelProperty("Image that this execution should be run in.")
+  private String image;
+  @ApiModelProperty("Command-line for this execution.")
+  private String program;
+  @ApiModelProperty("Container restrictions on this execution.")
+  private ContainerRestrictions restrictions;
+
+  @JsonCreator
+  public Execution(
+      @JsonProperty("image") String image,
+      @JsonProperty("program") String program,
+      @JsonProperty("restrictions") ContainerRestrictions restrictions) {
+    this.image = image;
+    this.program = program;
+    this.restrictions = restrictions;
+  }
+
+  public String getImage() {
+    return image;
+  }
+
+  public String getProgram() {
+    return program;
+  }
+
+  public ContainerRestrictions getRestrictions() {
+    return restrictions;
+  }
+
+  void setDefaultContainerRestriction(ContainerRestrictions defaultRestrictions) {
```

Comment:
I'd rather change this to a `withDefaultContainerRestrictions` and make new objects rather than mutate. The subclasses seem unclear later on; we're really just using them to thread information through the JSON decoder. There may be a more direct way to tell the JSON decoder the relevant default values; would you prefer that or a `with...` function?

Reply:
OK. Let's have a with... - I can't see what the issue is with subclasses but I'll trust you :-). Let's initialise the object with candidate default restrictions and then have a with... method for relaxing them to author restrictions. That seems safe and also means that field is never null.

Reply:
Subclasses: hmm, this was mostly gut feel, so I gave it a go, but then writing the `withProgram` function that I need later is not possible (I'd want to call `this.new` which isn't Java) unless I add in a builder pattern, and that is getting a bit ridiculous. So I did the with. Also, I used a marker for default rather than starting from candidate restrictions as it seemed wrong to make one of the default.